README.md
# deploy_war

Cookbook to make a war deploy/update in a big number of nodes.


CHANGELOG.md
##### Changelog v2.0.1 30/01/2017:

- New attribute `default["war"]["url_bbi"]` to manage BBI's deployments.

- Resource `log 'Eva.war downloaded.'` changed to `log "Eva.war downloaded from #{$war_url}"` in `deploy_proc.rb` recipe.

##### Changelog v2.0.0 24/01/2017:

- Global varibles `node_name`, `war_folder`, `username`, `password` and `war_url` were added to avoid redundance.

- Delete and replace some varibles to improve the cookbook's workflow.

- Delete `ruby_block Stop Eva (reverse)` resource in `reverse.rb` recipe to avoid unnecessary actions that cause Timeout errors.

- The last two dates of the changelog was corrected.

##### Changelog v1.2.7 20/01/2017:

- New attribute `default["war"]["url_panama"]` to manage Panamá's deployments.

- Varible `war_url` in `deploy_proc.rb` recipe stores the current url to download `Eva.war`.

##### Changelog v1.2.6 20/10/2016:

- Delete `ruby_block Stop Eva (manager)` resource in `deploy_proc.rb` recipe to avoid unnecessary actions that cause Timeout errors.

- Improve logic of `undeploy` function in `Eva` module.

##### Changelog v1.2.5 18/10/2016:

- Add new method in module `Tools` to fetch JSON data from an URL.

- Use `fetch` method in `ruby_block "Verify deployment"` resource of `deploy_proc.rb` and `reverse.rb` recipe to get data from `Eva/apilocalidad/version` service.

- Add varibles `node_name` and `prefix` in `deploy_proc.rb` and `reverse.rb` recipe to determine the company of shop updated when email is sent by `ruby_block "Verify deployment"` resource.

##### Changelog v1.2.4 11/10/2016:

- Add ruby_block resource in `prepare.rb` recipe to send a simple email when update process start in the nodes.

- `undeploy` change from method to function in module `Eva`, if undeploy Eva return true.

- Add resources to delete Eva.war and Eva folder in `deploy_proc.rb` recipe if undeploy fail for second time.


recipes/deploy.rb
# Cookbook Name:: deploy_war
# Recipe:: deploy
# Script to deploy a war file
# Copyright (c) 2016 The Authors, All Rights Reserved.

#------------ Variables -------------#
activeSessions = false
current_version = false
reachable = false
deploy = false
first_deploy = false
reverse = false

#----------- Main Process -----------#
# Ensure that war file is available to be downloaded
ruby_block 'Verify if war is available' do
  block do
    activeSessions = Tomcat.activeSessions?($username, $password)
    current_version = Eva.isCurrentVersion?($war_url, $username, $password)

    unless activeSessions || current_version
      if File.exist?("C:\\chef\\New_War\\Eva.war")
        Chef::Log.info("The war is available in folder C:\\chef\\New_War\\Eva.war to be deployed.")
      else
        if Tools.isReachable?($war_url)
          Chef::Log.info("The war is available to be downloaded.")
        else
          reachable = true
        end
      end
    end
  end
end

# Ruby code to do the deploy based on some decisions
ruby_block 'Deploy' do
  block do
    # Verify if there are active sessions or if the incoming war has the same version than the current
    unless activeSessions || current_version || reachable
      # if the war previously was deployed, it must be updated else deployed for first time.
      deploy = true
      unless Eva.status?($username, $password)[0]
        first_deploy = true
      end
      # Run deploy process
      run_context.include_recipe 'deploy_war::deploy_proc'
    else
      message = "Nothing to do at #{Time.now.iso8601}, something doesn't allow to continue with chef-client run:"
      message << "\n\- There are at least one open tomcat session." if activeSessions
      message << "\n\- The incoming war has the same version than the current (v#{$war_url[/\d+(.)\d+(.)\d+/]})." if current_version
      message << "\n\- The war is not available, something was wrong." if reachable
      Chef::Log.info(message)
      Tools.simple_email node["mail"]["to"], :message => message
    end
  end
end

# Reverse the Process (optional)
ruby_block 'Reverse the process' do
  block do
    unless Tools.isReachable?('http://localhost:8080/Eva/apilocalidad/version')
      unless first_deploy
        Chef::Log.warn("Something unexpected happened, reversing to the previous state.")
        reverse = true
        # Run reverse process
        run_context.include_recipe 'deploy_war::reverse'
      else
        message = "Something unexpected happened, Eva can not be deployed for first time."
        Chef::Log.fatal(message)
        Tools.send_email node["mail"]["to"], :message => message
      end
    end
  end
  only_if { deploy }
end

# Verify after reverse
ruby_block 'After reverse' do
  block do
    unless Tools.isReachable?('http://localhost:8080/Eva')
      message = "Something unexpected happened, Eva tried to reverse to its previous state but it failed."
      Chef::Log.fatal(message)
      Tools.send_email node["mail"]["to"], :message => message
    end
  end
  only_if { reverse }
end


recipes/prepare.rb
# Cookbook Name:: deploy_war
# Recipe:: prepare
# Preconfigurations
# Copyright (c) 2016 The Authors, All Rights Reserved.

#------------ logs ------------------#
log 'Service SQLServer started.' do
  action :nothing
end

#----------- Main Process -----------#
# Ruby code to send a simple email
ruby_block 'Report beginning of the update' do
  block do
    message = "Update process started at #{Time.now.iso8601}."
    Chef::Log.info(message)
    Tools.simple_email node["mail"]["to"], :message => message
  end
end

# Ensure that sqlserver is Running
powershell_script 'Start SQLSERVER' do
  code "start-service mssql*"
  notifies :write, 'log[Service SQLServer started.]', :immediately
  ignore_failure true
end
=begin
# Ruby code to pause chef-client in runtime while sql service start completely
ruby_block 'wait for sql service to start' do
  block do
    Tools.waitSQLSERVER
  end
=end

# Ensure that tomcat is Running
windows_service 'Tomcat7' do
  action :start
end

# Ruby code to pause chef-client in runtime while tomcat start completely
ruby_block 'wait for tomcat to start' do
  block do
    Tomcat.waitStart
  end
end


recipes/reverse.rb
# Cookbook Name:: deploy_war
# Recipe:: reverse
# Script to reverse the process if it fails
# Copyright (c) 2016 The Authors, All Rights Reserved.

#------------ Variables -------------#
reachable = false
prefix = ''

#----------- Main Process -----------#
ruby_block 'Verify deployment (reverse)' do
  block do
    reachable = Tools.isReachable?('http://localhost:8080/Eva/apilocalidad/version')

    if reachable
      f = Tools.fetch 'http://localhost:8080/Eva/apilocalidad/version'
      prefix = "B" if $node_name.include? "B"
      message = "Successful deployment, Eva v#{f["version"]} is ready in #{f["codLocalidad"]} #{f["descripLocalidad"]} !"
      Chef::Log.info(message)
      Tools.simple_email node["mail"]["to"], :message => message, :subject => "Chef Deployment on Node #{prefix}#{f["codLocalidad"]}"
    end
  end
end

ruby_block 'Undeploy Eva (reverse)' do
  block do
    Eva.undeploy($username, $password)
  end
  not_if { reachable }
end

# Stop Tomcat7
windows_service 'Stop Tomcat7 (reverse)' do
  service_name 'Tomcat7'
  action :stop
  not_if { reachable }
end

ruby_block 'wait while tomcat is stopped (reverse)' do
  block do
    Tomcat.waitStop
  end
  not_if { reachable }
end

# Execute command to copy current war into the folder previously created.
powershell_script 'Restore current war' do
  code "copy \'C:\\chef\\WarBackup\\Eva.war'  \'#{$war_folder}\'"
  not_if { reachable }
end

# Start Tomcat.
windows_service 'Start Tomcat7 (reverse)' do
  service_name 'Tomcat7'
  action :start
  not_if { reachable }
end

ruby_block 'wait for tomcat to start (reverse)' do
  block do
    Tomcat.waitStart
  end
  not_if { reachable }
end

ruby_block 'Wait while Eva is deployed (reverse)' do
  block do
    Eva.deploy($username, $password)
  end
  not_if { reachable }
end

ruby_block 'Verify reverse' do
  block do
    if Tools.isReachable?('http://localhost:8080/Eva')
      message = "Something unexpected happened, Eva reversed to its previous state."
      Chef::Log.info(message)
      Tools.send_email node["mail"]["to"], :message => message
    else
      Chef::Log.error("Request failed.")
    end
    Eva.expireSessions($username, $password)
  end
  not_if { reachable }
end


libraries/helpers.rb
include Chef::Mixin::PowershellOut
# Define general functions, methods, tools and utilities
module Tools
  require 'net/smtp'
  require 'json'
  require 'net/http'
  require 'mechanize'

  # Method to make web scraping and return body content
  def self.webScraping(url, username, password)
    agent = Mechanize.new
    agent.user_agent_alias = 'Windows Chrome'

    unless username.nil?
      agent.add_auth(url, username, password)
    end
    res = agent.get(url)
    return (res.body).to_s
  end

  # Function to know if one url is reachable
  def self.isReachable?(url)
    sw = true
    agent = Mechanize.new
    agent.user_agent_alias = 'Windows Chrome'
    agent.agent.http.verify_mode = OpenSSL::SSL::VERIFY_NONE
    tries = 3
    cont = 0

    begin
    	agent.read_timeout = 5 #set the agent time out
    	page = agent.get(url)
  	rescue
      cont += 1
      unless (tries -= 1).zero?
        Chef::Log.warn("Verifying if url #{url} is reachable (#{cont}/3) failed, try again in 1 minutes...")
        agent.shutdown
        agent = Mechanize.new { |agent| agent.user_agent_alias = 'Windows Chrome'}
        agent.request_headers
        sleep(60)
        retry
      else
        Chef::Log.error("The url #{url} isn't available.")
        sw = false
      end
    else
      sw = true
    ensure
      agent.history.pop()   #delete this request in the history
    end

    return sw
  end

  # Method to fetch data in JSON format from an URL
  def self.fetch(url)
    resp = Net::HTTP.get_response(URI.parse(url))
    data = resp.body
    return JSON.parse(data)
  end

  # Method to start SQLSERVER
  def self.waitSQLSERVER
    connectable = false
    3.times do
      sql = powershell_out!("Invoke-Sqlcmd -Query \'Select name From sys.databases\'")

      if sql.stdout[/error/].nil?
        connectable = true
        Chef::Log.info("SQLSERVER Started.")
        break
      else
        Chef::Log.info("Waiting 20 seconds for SQLSERVER to continue...")
        sleep(20)
      end
    end

    unless connectable
      Chef::Log.error("Could not connect to the SQLSERVER database.")
    end
  end

  def self.unindent string
    first = string[/\A\s*/]
    string.gsub /^#{first}/, ''
  end

  def self.send_email(to ,opts={})
    opts[:server]      ||= 'smtp.office365.com'
    opts[:port]        ||= 587
    opts[:from]        ||= 'barcoder@redsis.com'
    opts[:password]    ||= 'Orion2015'
    opts[:from_alias]  ||= 'Chef Reporter'
    opts[:subject]     ||= "Chef Deployment on Node #{Chef.run_context.node.name}"
    opts[:message]     ||= "..."

    filename = "C:\\chef\\log-#{Chef.run_context.node.name}"
    # Read a file and encode it into base64 format
    encodedcontent = [File.read(filename)].pack("m")   # base64

    marker = "AUNIQUEMARKER"

    # Define the main headers.
    header = <<-HEADER
      From: #{opts[:from_alias]} <#{opts[:from]}>
      To: <#{to}>
      Subject: #{opts[:subject]}
      MIME-Version: 1.0
      Content-Type: multipart/mixed; boundary=#{marker}
      --#{marker}
    HEADER

    # Define the message action
    body = <<-BODY
      Content-Type: text/plain
      Content-Transfer-Encoding:8bit

      #{opts[:message]}
      --#{marker}
    BODY

    # Define the attachment section
    attached = <<-ATTACHED
      Content-Type: multipart/mixed; name=\"#{filename}\"
      Content-Transfer-Encoding:base64
      Content-Disposition: attachment; filename="#{filename}"

      #{encodedcontent}
      --#{marker}--
    ATTACHED

    mailtext = unindent header + body + attached

    smtp = Net::SMTP.new(opts[:server], opts[:port])
    smtp.enable_starttls_auto
    smtp.start(opts[:server], opts[:from], opts[:password], :login)
    smtp.send_message(mailtext, opts[:from], to)
    smtp.finish
  end

  def self.simple_email(to, opts={})
    opts[:server]      ||= 'smtp.office365.com'
    opts[:port]        ||= 587
    opts[:from]        ||= 'barcoder@redsis.com'
    opts[:password]    ||= 'Orion2015'
    opts[:from_alias]  ||= 'Chef Reporter'
    opts[:subject]     ||= "Chef Start on Node #{Chef.run_context.node.name}"
    opts[:message]     ||= "..."

    message = <<-MESSAGE
      From: #{opts[:from_alias]} <#{opts[:from]}>
      To: <#{to}>
      Subject: #{opts[:subject]}

      #{opts[:message]}
    MESSAGE

    mailtext = unindent message

    smtp = Net::SMTP.new(opts[:server], opts[:port])
    smtp.enable_starttls_auto
    smtp.start(opts[:server], opts[:from], opts[:password], :login)
    smtp.send_message(mailtext, opts[:from], to)
    smtp.finish
  end
end

# Define functions, methods, tools and utilities to work with Tomcat
module Tomcat
  require 'mechanize'

  # Function to get the war folder
  def self.getWarFolder
    if File.directory?("C:\\Program Files (x86)\\Apache Software Foundation\\Tomcat 7.0\\webapps")
      return "C:\\Program Files (x86)\\Apache Software Foundation\\Tomcat 7.0\\webapps"
    else
      return "C:\\Program Files\\Apache Software Foundation\\Tomcat 7.0\\webapps"
    end
  end

  # Function to know if tomcat is Running
  def self.isRunning?
    tomcat = powershell_out!("(Get-Service Tomcat7).Status -eq \'Running\'")

    if tomcat.stdout[/True/]
      return true
    else
      return false
    end
  end

  # Function to know if tomcat is Stop
  def self.isStop?
    tomcat = powershell_out!("(Get-Service Tomcat7).Status -eq \'Stopped\'")

    if tomcat.stdout[/True/]
      return true
    else
      return false
    end
  end

  # Method to pause execution while tomcat start
  def self.waitStart
    if isRunning?
      agent = Mechanize.new
      agent.user_agent_alias = 'Windows Chrome'

    	begin
      	agent.read_timeout = 5 #set the agent time out
      	page = agent.get('http://localhost:8080')
      	agent.history.pop()   #delete this request in the history
        Chef::Log.info("Tomcat7 Started")
    	rescue
    		Chef::Log.info("Waiting 2.5 minutes for Tomcat7 to continue...")
    		agent.shutdown
    		agent = Mechanize.new { |agent| agent.user_agent_alias = 'Windows Chrome'}
    		agent.request_headers
    		sleep(150)
    		retry
    	end
    end
  end

  # Method to pause execution while tomcat stop
  def self.waitStop
    while !isStop? do
      Chef::Log.info("Waiting 2 minutes for Tomcat7 to continue...")
      sleep(120)
    end
    Chef::Log.info("Tomcat7 stopped !")
  end

  # Method to get the list of applications deployed and its number of sessions
  def self.sessionList(username, password)
    session_mesage = Tools.webScraping('http://localhost:8080/manager/text/list', username, password)
    sessions = Array.new

    session_mesage.each_line do |line|
      session = line.split(':')
      sessions.push(session)
    end
    return sessions
  end

  # Function to validate if there are active sessions in tomcat manager
  def self.activeSessions?(username, password)
    if isRunning?
      sessions = sessionList(username, password)
      i = 0
      sw = false
      while (i < sessions.length && !sw)
        session = sessions[i]
        if (session[0].start_with?("/") && !session[0].eql?("/manager") && session[2].to_i > 0)
          sw = true
        end
        i += 1
      end
      return sw
    else
      return false
    end
  end
end

# Define functions, methods, tools and utilities to work with Eva
module Eva
  require 'mechanize'

  # Function to know if Eva is Running
  def self.status?(username, password)
    if Tomcat.isRunning?
      sessions = Tomcat.sessionList(username, password)
      i = 0
      exist = false
      status = ""
      while (i < sessions.length && !exist)
        session = sessions[i]
        if (session[0].eql?("/Eva"))
          exist = true
          status = session[1]
        end
        i += 1
      end
      return [exist, status]
    else
      return [false, ""]
    end
  end

  # Method to start Eva
  def self.start(username, password)
    # Verify if Eva exist and it is sttoped
    if status?(username, password)[0] && status?(username, password)[1].eql?("stopped")
      agent = Mechanize.new
      agent.user_agent_alias = 'Windows Chrome'
      agent.add_auth('http://localhost:8080/manager/text/start?path=/Eva', username, password)
      page = agent.get('http://localhost:8080/manager/text/start?path=/Eva')

      while status?(username, password)[1].eql?("stopped") do
        Chef::Log.info("Waiting 2 minutes while starting Eva to continue...")
        sleep(120)
      end

      Chef::Log.info("Eva Started !")
    end
  end

  # Method to stop Eva
  def self.stop(username, password)
    if status?(username, password)[0] && status?(username, password)[1].eql?("running")
      agent = Mechanize.new
      agent.user_agent_alias = 'Windows Chrome'
      agent.add_auth('http://localhost:8080/manager/text/stop?path=/Eva', username, password)
      page = agent.get('http://localhost:8080/manager/text/stop?path=/Eva')

      while status?(username, password)[1].eql?("running") do
        Chef::Log.info("Waiting 2 minutes while stopping Eva to continue...")
        sleep(120)
      end

      Chef::Log.info("Eva Stopped !")
    end
  end

  # Method to undeploy Eva
  def self.undeploy(username, password)
    if status?(username, password)[0]
      sw = true
      agent = Mechanize.new
      agent.user_agent_alias = 'Windows Chrome'
      tries = 3
      cont = 0

      begin
        agent.add_auth('http://localhost:8080/manager/text/undeploy?path=/Eva', username, password)
        page = agent.get('http://localhost:8080/manager/text/undeploy?path=/Eva')
      rescue
        cont += 1
        unless (tries -= 1).zero?
          Chef::Log.warn("Undeploy Eva (#{cont}/3) failed, try again in 1 minutes...")
          agent.shutdown
          agent = Mechanize.new { |agent| agent.user_agent_alias = 'Windows Chrome'}
          agent.request_headers
          sleep(60)
          retry
        else
          Chef::Log.error("Could not execute undeploy Eva.")
          sw = false
        end
      else
        sw = true
      ensure
        agent.history.pop()   #delete this request in the history
      end

      if sw
        i = 0
        while status?(username, password)[0] && i < 5 do
          Chef::Log.info("Waiting 2 minutes while undeploying Eva to continue...")
          i += 1
          sleep(120)
        end

        if status?(username, password)[0]
          agent.shutdown
          sleep(60)
      		agent = Mechanize.new { |agent| agent.user_agent_alias = 'Windows Chrome'}

          begin
            agent.add_auth('http://localhost:8080/manager/text/undeploy?path=/Eva', username, password)
            page = agent.get('http://localhost:8080/manager/text/undeploy?path=/Eva')
          rescue
            Tools.simple_email 'cbeleno@redsis.com', :message => 'Needing a manual adjustment !', :subject => "Chef Undeploy on Node #{Chef.run_context.node.name}"
          end

          i = 0
          while status?(username, password)[0] && i < 5 do
            Chef::Log.info("Waiting 2 minutes more while undeploying Eva to continue...")
            i += 1
            sleep(120)
          end
        end

        unless status?(username, password)[0]
          Chef::Log.info("Eva Undeployed !")
          return true
        else
          Chef::Log.info("Could not undeploy Eva.")
          return false
        end

      else
        return false
      end

    end
  end

  # Method to deploy Eva
  def self.deploy(username, password)
    while !status?(username, password)[0] &&  !status?(username, password)[1].eql?("running") do
      Chef::Log.info("Waiting 2 minutes for Eva to continue...")
      sleep(120)
    end

    Chef::Log.info("Eva Deployed !")
    Chef::Log.info("Waiting 2 minutes more to continue...")
    sleep(120)
  end

  # Function to validate if the version of new war is the same than the current
  def self.isCurrentVersion?(war_url, username, password)
  	sw = false
  	if status?(username, password)[0]
      if Tools.isReachable?('http://localhost:8080/Eva/apilocalidad/version')
    		version = Tools.webScraping('http://localhost:8080/Eva/apilocalidad/version', nil, nil)
    		unless version[/\d+(\.)\d+(\.)\d+/].nil?
    			current_version = version[/\d+(.)\d+(.)\d+/]
    			new_version = war_url[/\d+(.)\d+(.)\d+/]
    			sw = current_version.eql?(new_version)
    		end
      else
        Chef::Log.warn("Could not determine the version of Eva.")
      end
  	end
  	return sw
  end

  def self.expireSessions(username, password)
    agent = Mechanize.new
    agent.user_agent_alias = 'Windows Chrome'
    agent.add_auth('http://localhost:8080/manager/text/expire?path=/Eva&idle=0', username, password)
    page = agent.get('http://localhost:8080/manager/text/expire?path=/Eva&idle=0')
  end
end

Chef::Recipe.send(:include, Tools)
Chef::Recipe.send(:include, Tomcat)
Chef::Recipe.send(:include, Eva)


recipes/deploy_proc.rb
# Cookbook Name:: deploy_war
# Recipe:: deploy_proc
# Script with deploy process
# Copyright (c) 2016 The Authors, All Rights Reserved.

#------------ Variables -------------#
first_time = true
undeploy = true
prefix = ''

#----------- Main Process -----------#
log "Eva.war downloaded from #{$war_url}" do
  action :nothing
end

# Verify if Eva is deployed already
ruby_block 'Verify if Eva is deployed' do
  block do
    if Eva.status?($username, $password)[0]
      first_time = false
      Chef::Log.info("Eva will be updated.")
    else
      Chef::Log.info("Eva will be deployed for the first time.")
    end
  end
end

# Create backupWar folder to store war files replaced.
directory "C:\\chef\\WarBackup" do
  not_if { File.directory?("C:\\chef\\WarBackup") || first_time }
end

# Delete the war previously was replaced
file "C:\\chef\\WarBackup\\Eva.war" do
  action :delete
  only_if { File.exist?("C:\\chef\\WarBackup\\Eva.war") }
end

# Execute command to copy current war into the folder previously created.
powershell_script 'Store current war' do
  code "copy \'#{$war_folder}\\Eva.war\' \'C:\\chef\\WarBackup\'"
  only_if { File.directory?("C:\\chef\\WarBackup") && File.exist?("#{$war_folder}\\Eva.war") }
end

ruby_block 'Undeploy Eva (Manager)' do
  block do
    undeploy = Eva.undeploy($username, $password)
  end
end

# Stop Tomcat7
windows_service 'Stop Tomcat7 (deploy)' do
  service_name 'Tomcat7'
  action :stop
end

ruby_block 'wait while tomcat is stopped' do
  block do
    Tomcat.waitStop
  end
end

# Delete current war.
file "#{$war_folder}\\Eva.war" do
  action :delete
  only_if { File.exist?("#{$war_folder}\\Eva.war") && !undeploy }
end

# Delete Eva's folder.
directory "#{$war_folder}\\Eva" do
  recursive true
  action :delete
  only_if { File.directory?("#{$war_folder}\\Eva") && !undeploy }
end

# Execute command to copy new war into the webapps folder
powershell_script 'Copy new war' do
  code "copy \'C:\\chef\\New_War\\Eva.war\' \'#{$war_folder}\'"
  only_if { File.exist?("C:\\chef\\New_War\\Eva.war") }
end

# Download the war from remote source if it's necessary.
remote_file 'Download Eva.war' do
  path "#{$war_folder}\\Eva.war"
  source $war_url
  notifies :write, "log[Eva.war downloaded from #{$war_url}]", :immediately
  not_if { File.exist?("C:\\chef\\New_War\\Eva.war") }
end

# Start Tomcat.
windows_service 'Start Tomcat7 (deploy)' do
  service_name 'Tomcat7'
  action :start
end

# Ruby code to pause chef-client in runtime while tomcat start completely
ruby_block 'wait for tomcat to start (deploy)' do
  block do
    Tomcat.waitStart
  end
end

# Ruby code to pause chef-client in runtime while Eva deploy completely
ruby_block 'Wait while Eva is deployed' do
  block do
    Eva.deploy($username, $password)
  end
end

# Ruby code to verify the deployment of Eva
ruby_block 'Verify deployment' do
  block do
    if Tools.isReachable?('http://localhost:8080/Eva/apilocalidad/version')
      f = Tools.fetch 'http://localhost:8080/Eva/apilocalidad/version'
      prefix = "B" if $node_name.start_with? "B"
      prefix = "P" if $node_name.start_with? "P"
      message = "Successful deployment, Eva v#{f["version"]} is ready in #{f["codLocalidad"]} #{f["descripLocalidad"]} !"
      Chef::Log.info(message)
      Tools.simple_email node["mail"]["to"], :message => message, :subject => "Chef Deployment on Node #{prefix}#{f["codLocalidad"]}"
    else
      Chef::Log.error("Request failed.")
    end
  end
end


recipes/default.rb
# Cookbook Name:: deploy_war
# Recipe:: default
# Script to decide if deploy or update a war file
# Copyright (c) 2016 The Authors, All Rights Reserved.

#------------ Variables -------------#
$node_name = Chef.run_context.node.name.to_s
$war_folder = Tomcat.getWarFolder
$username = node["tomcat-manager"]["username"]
$password = node["tomcat-manager"]["password"]
$war_url =
  if $node_name.start_with?('B') || $node_name.start_with?('QA') #|| $node_name.start_with?('RD')
    node["war"]["url_bbi"]
  elsif $node_name.start_with? 'P'
    node["war"]["url_panama"]
  else
    node["war"]["url"]
  end
shops = %w(
  100 101 102 103 104 105 106 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 124 125 126 127 128 129 131 134
  135 136 137 139 140 142 143 144 146 147 148 149 151 152 153 156 157 160 161 162 164 165 166 167 168 169 170 171 150 154
)

#------------ Main process -------------#
file 'C:\chef\New_War\Eva.war' do
  action :delete
  only_if { shops.include?($node_name) }
end

# System preconfigurations
include_recipe 'deploy_war::prepare'

# Deploy
include_recipe 'deploy_war::deploy'
